src/inference/src/dev/make_tensor.cpp

```diff
@@ -182,43 +182,36 @@ class AllocatedTensor : public ViewTensor {
                          OPENVINO_ASSERT(allocator, "Allocator was not initialized");
                          auto num_elements = shape_size(shape);
                          auto data = const_cast<Allocator&>(allocator).allocate(element_type.size() * num_elements);
-                         init(data, element_type, shape);
+                         initialize_elements(data, element_type, shape);
                          return data;
                      }()},
           m_allocator{allocator} {}
 
     ~AllocatedTensor() {
-        auto num_elements = get_size();
-        destroy(0, num_elements);
-        m_allocator.deallocate(m_ptr, get_byte_size());
+        destroy_memory();
     }
 
     void set_shape(ov::Shape new_shape) override {
         if (m_shape == new_shape)
             return;
 
-        auto old_num_elements = get_size();
-        auto old_byte_size = get_byte_size();
         m_shape = std::move(new_shape);
-        auto new_num_elements = get_size();
 
-        if (get_byte_size() > old_byte_size) {
+        if (get_size() > get_capacity()) {
```

Comment:
here I assume that when a memory is allocated, it's immediately initialized within `initialize_elements` function. In this case we don't really need to care about multiple cases and track capacity 
Do you consider such implementation as suboptimal? 

When tensor is initially allocated, we need to initialize all elements with values (in case of complex objects like `std::string`). Later:
- When user calls `set_shape` with size bigger than capacity => no field for optimization, because we need to remove objects and deallocate memory
- When `set_shape` is called with size `current_size < size < capacity`, we need to call default ctor.
- When `set_shape` is called with smaller size than current `size`, we could remove some objects. But what if on next iterations `set_shape` is called with size from step 2? We need to call default ctors again. So, we potentially removing / creating multiple string objects many times. What I'm doing here - defer final deinitialization of string objects till `destroy_memory` call. Do you consider it as a problem?

Reply:
Consider add comment that this check is to avoid calls of ctors/dtors for allocated complex type if not required.